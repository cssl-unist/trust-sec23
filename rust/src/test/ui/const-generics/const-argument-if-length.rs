// revisions: full min

#![cfg_attr(full, allow(incomplete_features))]
#![cfg_attr(full, feature(const_generics))]
#![cfg_attr(min, feature(min_const_generics))]

pub const fn is_zst<T: ?Sized>() -> usize {
    if std::mem::size_of::<T>() == 0 {
        //[full]~^ ERROR the size for values of type `T` cannot be known at compilation time
        1
    } else {
        0
    }
}

pub struct AtLeastByte<T: ?Sized> {
    value: T,
    //~^ ERROR the size for values of type `T` cannot be known at compilation time
    pad: [u8; is_zst::<T>()],
    //[min]~^ ERROR generic parameters may not be used in const operations
    //[full]~^^ ERROR evaluation of constant value failed
}

fn main() {}
