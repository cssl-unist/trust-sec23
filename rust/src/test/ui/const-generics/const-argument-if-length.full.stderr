error[E0277]: the size for values of type `T` cannot be known at compilation time
  --> $DIR/const-argument-if-length.rs:8:28
   |
LL | pub const fn is_zst<T: ?Sized>() -> usize {
   |                     - this type parameter needs to be `Sized`
LL |     if std::mem::size_of::<T>() == 0 {
   |                            ^ doesn't have a size known at compile-time
   | 
  ::: $SRC_DIR/core/src/mem/mod.rs:LL:COL
   |
LL | pub const fn size_of<T>() -> usize {
   |                      - required by this bound in `std::mem::size_of`

error[E0080]: evaluation of constant value failed
  --> $DIR/const-argument-if-length.rs:19:15
   |
LL |     pad: [u8; is_zst::<T>()],
   |               ^^^^^^^^^^^^^ referenced constant has errors

error[E0277]: the size for values of type `T` cannot be known at compilation time
  --> $DIR/const-argument-if-length.rs:17:12
   |
LL | pub struct AtLeastByte<T: ?Sized> {
   |                        - this type parameter needs to be `Sized`
LL |     value: T,
   |            ^ doesn't have a size known at compile-time
   |
   = note: only the last field of a struct may have a dynamically sized type
   = help: change the field's type to have a statically known size
help: borrowed types always have a statically known size
   |
LL |     value: &T,
   |            ^
help: the `Box` type always has a statically known size and allocates its contents in the heap
   |
LL |     value: Box<T>,
   |            ^^^^ ^

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0080, E0277.
For more information about an error, try `rustc --explain E0080`.
